components/html-block/html-block.js

```diff
@@ -129,22 +144,18 @@ class HtmlBlock extends LitElement {
 			const fragment = template ? document.importNode(template.content, true) : null;
 			if (fragment) {
 
-				const hasMathML = !!fragment.querySelector('math');
-
-				const temp = document.createElement('div');
+				let temp = document.createElement('div');
 				temp.appendChild(fragment);
-				const fragmentHTML = temp.innerHTML;
-
-				const isLatexSupported = window.D2L && window.D2L.LP && window.D2L.LP.Web.UI.Flags.Flag('us125413-mathjax-render-latex', true);
-				if (hasMathML || (isLatexSupported && /\$\$|\\\(/.test(fragmentHTML))) {
-					const mathJaxConfig = { renderLatex: isLatexSupported };
-					await loadMathJax(mathJaxConfig);
-
-					this._renderContainer.innerHTML = `<mjx-doc><mjx-head></mjx-head><mjx-body>${fragmentHTML}</mjx-body></mjx-doc>`;
-					window.MathJax.typesetShadow(this.shadowRoot);
-				} else {
-					this._renderContainer.innerHTML = fragmentHTML;
-				}
+
+				// parallelize prepare (loading dependencies)
```

Comment:
The current approach iterates over each of the renderers, loads the dependencies, and invokes their render methods if necessary. The one thing I don't like about this approach is that each renderer ends up searching the HTML.

An alternative approach that I have considered is to search once for fragments that should be replaced with something like `elem.querySelectorAll('[data-d2l-html-block-renderer="renderer key"]')`.  For each element returned, request the required renderer using `requestInstance`, and the proceed with loading dependencies and doing the replacements.  The problem with this approach is - it will not work for math, where we need to query for `math`, `\(`, `\[`, `$$` instead of a common attribute like `data-d2l-html-block-renderer`.

Reply:
Interesting. Like the ideal case is that if there's no fragment for a particular renderer, no work gets done and nothing gets loaded. The way you have it here, `prepare()` always gets called for all renderers, which is going to result in an API call for the mentions stuff even if there are no mentions.

I don't think I'd be too worried about the multiple `querySelectorAll` calls -- they're pretty quick. Could `prepare` and `render` be combined, making `render` async? That way the plugins could query the HTML and bail out if it doesn't contain anything of interest. If it does contain something, then they could do a 1-time "prepare" to load any dependencies and then finally do the actual replacement.

Reply:
> The way you have it here, `prepare()` always gets called for all renderers, which is going to result in an API call for the mentions stuff even if there are no mentions.

That's a good point.  The @mentions `prepare` could query for the attribute and return `false` before doing the API check.
 
> Could `prepare` and `render` be combined, making `render` async? That way the plugins could query the HTML and bail out if it doesn't contain anything of interest. If it does contain something, then they could do a 1-time "prepare" to load any dependencies and then finally do the actual replacement.

I was originally thinking along those lines too but then thought there would be issues due to (the math render in particular) returning a different outer element.  I think this may be possible as long as the outer element (`div` passed into each `render` call) is exactly the same, which is the same element ultimately referenced on line 158. I don't think it should matter, but the order they run could become inconsistent.

Reply:
I guess I'm just questioning whether having a separate `prepare()` makes sense given that it may need to look at the HTML anyway (so it would need to be passed the node) to see if it even needs to do anything. So if you're passing the node to both `prepare()` and `render()`, why not just combine them and for plugins that need to do some kind of preparation (import stuff, do API calls) they can be smart enough to just do that work the first time.

Reply:
I wonder about the cost of forcing the prep work to be synchronous (if it's folded into `render()`). I know those query selectors are fast, but I'm less sure about the regex the math stuff needs to test, especially on what could potentially be very large HTML files (potentially MB in size). I guess it doesn't matter, since the rendering can't happen until the prep is done and the user's browser is going to incur that time/processing cost regardless. In those cases, I imagine perhaps the math `prepare()` could end up dwarfing the other providers and it wouldn't cost much extra time to make it synchronous. :shrug:

Reply:
Yeah I agree that doing it on one if possible would be ideal.  Keeping them async requires that the top-level element reference doesn't change though (which the math renderer currently does replace).  I am going to try and rejig thing slightly to avoid that.

Reply:
Ok, I've thought about this some more and I think my original concern is legit.  Take these two renderers:

```javascript
// render 1
async render1(elem) {
    const stuffToReplace = elem.querySelector('[some-attr]');
    if (!stuffToReplace) return;

    await loadStuffDependencies();

    stuffToReplace.innerHTML = '...' / stuffToReplace.replaceChild(...);
}

//render 2 (akin to what mathjax does)
async render2(elem) {
    const fooToReplace = elem.querySelector('some-elem');
    if (!fooToReplace) return;

    await loadFooDependencies();

    elem.innerHTML = fooProcessor(elem.innerHTML);
}

await Promise.all(
    render1(elem),
    render2(elem)
);
```
This issue is that the replacement may be done in any order, and one replacement may invalidate node references within another render's closure.  In the above example, if `render2` dependencies are loaded first, its replacement would run first and completely invalidate `stuffToReplace`.  To some extent, this can be mitigated by not using `innerHTML`, but I think there is still opportunity of one replacement to invalidate node references in another render call.

From the beginning, I was trying to be optimistic about loading the dependencies (in parallel), perhaps to the peril of the code at hand, when ultimately, the only scenario when it actually matters is when a user would be seeing rich text containing these things all together, and for the very first time (uncached).  So perhaps we just make these calls synchronous, which avoid this issue altogether, and we can do it all in one method safely.

Reply:
Right, I see the concern now. I'm less worried about the dependency loading step (which is cacheable like you said) and more concerned about that API call to check if the user has permission. That's likely _not_ being cached, so would literally run once per page load... and it's likely not fast either, probably 100-200ms.

So bailing early (doing the querySelector and if there's nothing, bailing) will be key here.

Reply:
Ok, I've updated to make this process rely on a single method per "plugin".

Re the API call that @mentions uses - maybe we can optimize beyond just checking for @mentions as per my [comment below](https://github.com/BrightspaceUI/core/pull/1282#discussion_r618832761).  I have added the short-circuit however.